clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -376,39 +376,58 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
         // by markers.
         bool have_cond_branch_target = false;
         addr_t cond_branch_target = 0;
+        bool saw_repeated_syscall_instrs_with_same_pc = false;
         if (shard->prev_instr_.instr.addr != 0 /*first*/ &&
-            type_is_instr_direct_branch(shard->prev_instr_.instr.type) &&
             // We do not bother to support legacy traces without encodings.
             TESTANY(OFFLINE_FILE_TYPE_ENCODINGS, shard->file_type_)) {
             addr_t trace_pc = shard->prev_instr_.instr.addr;
-            if (shard->prev_instr_.instr.encoding_is_new)
-                shard->branch_target_cache.erase(trace_pc);
-            auto cached = shard->branch_target_cache.find(trace_pc);
-            if (cached != shard->branch_target_cache.end()) {
-                have_cond_branch_target = true;
-                cond_branch_target = cached->second;
-            } else {
-                instr_t instr;
-                instr_init(GLOBAL_DCONTEXT, &instr);
-                const app_pc decode_pc =
-                    const_cast<app_pc>(shard->prev_instr_.instr.encoding);
-                const app_pc next_pc =
-                    decode_from_copy(GLOBAL_DCONTEXT, decode_pc,
-                                     reinterpret_cast<app_pc>(trace_pc), &instr);
-                if (next_pc == nullptr || !opnd_is_pc(instr_get_target(&instr))) {
-                    // Neither condition should happen but they could on an invalid
-                    // encoding from raw2trace or the reader so we report an
-                    // invariant rather than asserting.
-                    report_if_false(shard, false, "Branch target is not decodeable");
-                } else {
+            instr_t instr;
+            instr_init(GLOBAL_DCONTEXT, &instr);
```

Comment:
So instr is really the previous instr?

Reply:
Right and that's one of the tricky aspects of the decoding logic: We won't be able to know if the "current" instr was a syscall until we process the instr after it. But by the time we get to process the following instr (and discover instr was a syscall), we will have already hit a generic PC discontinuity.